diff --git a/Android.bp b/Android.bp
index 46954cf..fd15c12 100644
--- a/Android.bp
+++ b/Android.bp
@@ -95,6 +95,7 @@ cc_library {
         "//external/rust/android-crates-io/crates/libbpf-rs",
         "//external/rust/android-crates-io/crates/libbpf-sys",
         "//external/stg",
+	"//external/agent",
         "//hardware/interfaces/health/utils/libhealthloop", // For use in tests only.
         // Because libbpf depends on the GPL-licensed libelf, its use should be restricted to the
         // standalone bpf loader binary. This visibility must not be widened.
diff --git a/src/bpf_helpers.h b/src/bpf_helpers.h
index cd17f6d..f16caaa 100644
--- a/src/bpf_helpers.h
+++ b/src/bpf_helpers.h
@@ -15,6 +15,11 @@
 #define __array(name, val) typeof(val) *name[]
 #define __ulong(name, val) enum { ___bpf_concat(__unique_value, __COUNTER__) = val } name
 
+//sifter
+static int (*bpf_probe_read_sleepable)(void* dst, int size, void* unsafe_ptr) = (void*) BPF_FUNC_probe_read_sleepable;
+static int (*bpf_check_fd)(char* path, int fd) = (void*) BPF_FUNC_check_fd;
+
+
 /*
  * Helper macro to place programs, maps, license in
  * different sections in elf_bpf file. Section names
